### DISCUSSION TITLE
Replacing bcrypt

### DISCUSSION THREAD

**Discussion by jezzzm** [keystonejs/keystone#7568] [Questions] [answered]
hey everyone, looking for some motherhashing help!

i’m moving an old wordpress site to keystone and hoping to not require users to set a new password. I’ve found a [nodejs library](https://github.com/glauberportella/password-hash) which is a port of the wp password hashing method. have verified with my own password + hash that it does work.

i have seeded the database with the hashed user passwords from wp using `prisma.user.createMany` instead of `db.User.createMany` as i understand the former doesn’t hash it again?

next i replaced the default bcrypt with my little wrapper over node-phpass to match the bcrypt interface:

```ts
import { PasswordHash } from 'node-phpass';

const LENGTH = X;
const PORTABLE = true;
const PHPVERSION = Y;

const hasher = new PasswordHash(LENGTH, PORTABLE, PHPVERSION);

export const hash = async (plain: string): Promise<string> => {
  return hasher.HashPassword(plain);
};

export const compare = async (plain: string, hash: string): Promise<boolean> => {
  return hasher.CheckPassword(plain, hash);
};
```

and added to user schema:
```ts
import { compare, hash } from '../lib/wp-password-hash';
//...
    password: password({
      validation: { isRequired: true },
      access: {
        update: isPerson,
        read: isPerson
      },
      ui: {
        createView: { fieldMode: 'hidden' },
        itemView: { fieldMode: 'hidden' },
        listView: { fieldMode: 'hidden' }
      },
      bcrypt: { // ts happy with this interface
        compare,
        hash
      }
    }),
```

all’s good so far. but! i’m running into a problem with `validateSecret `, identity protection is kicking in and passing a garbage secret (`'simulated-password-to-counter-timing-attack'`) to the compare func. i can’t figure out why…

any help would be much appreciated!

**Comment by dcousens** [comment, marked as answer]
Hi @jezzzm,

Undoubtedly we were hoping that the `'simulated-password-to-counter-timing-attack'` string might have provided context as to what was happening here,  but for posterity I'll describe below the idea generally.

A [timing attack](https://en.wikipedia.org/wiki/Timing_attack) is something a nefarious actor can do to learn information about your system by measuring the timing of particular operations. 

This string is provided to the `compare` function to specifically _fail_, but none-the-less approximately be equivalent in timing to when a user actually exists.  If you don't do this, there is the possibility that an attacker can measure the timing of the request to determine if a user exists within the database, or not.
